[JIT] Add a pass for removing a first (self) argument from a graph if it is unused.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57074 [TensorExpr] Add methods for inspecting generated code in `TensorExprKernel`.
* #57382 [TensorExpr] Remove `dtype_` and add `buf_` fields to `CodeGen::BufferArg`.
* **#57169 [JIT] Add a pass for removing a first (self) argument from a graph if it is unused.**
* #57076 [JIT] Add a pass for annotating graph with input types derived from sample inputs.

The pass is planned to be used in AOT pipeline, where we expect input
graphs to be functional. As such, these graphs should not use 'self'
argument even if it is present, and thus it can be remove safely.

Differential Revision: [D28128328](https://our.internmc.facebook.com/intern/diff/D28128328)

Commit:
[JIT] Add a pass for removing a first (self) argument from a graph if it is unused.

The pass is planned to be used in AOT pipeline, where we expect input
graphs to be functional. As such, these graphs should not use 'self'
argument even if it is present, and thus it can be remove safely.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add a pass for removing a first (self) argument from a graph if it is unused."


The pass is planned to be used in AOT pipeline, where we expect input
graphs to be functional. As such, these graphs should not use 'self'
argument even if it is present, and thus it can be remove safely.

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -239,6 +239,21 @@ void annotateInputShapes(
+  if (graph->inputs().size() == 0) {
+    return graph;
+  }
+  jit::Value* self_argument = graph->inputs().at(0);
```
Kind of a dumb question but how do we know this is a `self` argument?  Or is that kind of a red herring, and we just always try to remove an unused first argument (which happens to usually be self)?

ZolotukhinM replied:
We check that the type of the argument is "module" in a check below, but overall it is possible that this argument is not `self`. The assumption here is that this pass will be used when having a self arguments prevents whole-graph compilation from happening, e.g.:
```
aot_artifact = torch.jit._C.compile_graph(graph)
# complains about unsupported type `something.something.Module`
graph2 = torch.jit._C.remove_unused_self(graph)
aot_artifact = torch.jit._C.compile_graph(graph2)
# succeeds
```

bertmaher reviewed (approved):
I have a question for my own understanding but the code looks fine

Commit:
Update on "[JIT] Add a pass for removing a first (self) argument from a graph if it is unused."


The pass is planned to be used in AOT pipeline, where we expect input
graphs to be functional. As such, these graphs should not use 'self'
argument even if it is present, and thus it can be remove safely.

Differential Revision: [D28128328](https://our.internmc.facebook.com/intern/diff/D28128328)

[ghstack-poisoned]
